elf_reader.go

```diff
@@ -610,28 +657,70 @@ func mapSpecFromBTF(spec *btf.Spec, name string) (*MapSpec, error) {
 			}
 
 		case "pinning":
+			if inner {
+				return nil, errors.New("inner maps can't be pinned")
+			}
+
 			pinning, err := uintFromBTF(member.Type)
 			if err != nil {
 				return nil, fmt.Errorf("can't get pinning: %w", err)
 			}
 
 			pinType = PinType(pinning)
 
-		case "key", "value":
+		case "values":
+			// The 'values' field in BTF map definitions is used for declaring map
+			// value types that are references to other BPF objects, like other maps
+			// or programs. It is always expected to be an array of pointers.
+			if i != len(members)-1 {
+				return nil, errors.New("'values' must be the last member in a BTF map definition")
+			}
+
+			valueType, err := resolveBTFArrayMacro(member.Type)
+			if err != nil {
+				return nil, fmt.Errorf("can't resolve type of member 'values': %w", err)
+			}
+
+			switch t := valueType.(type) {
+			case *btf.Struct:
+				// The values member pointing to an array of structs means we're expecting
+				// a map-in-map declaration.
+				if MapType(mapType) != ArrayOfMaps && MapType(mapType) != HashOfMaps {
+					return nil, errors.New("outer map needs to be an array or a hash of maps")
+				}
+				if inner {
+					return nil, fmt.Errorf("nested inner maps are not supported")
+				}
+
+				// Inner maps are always unnamed, so provide an empty name value.
```

Comment:
Just as an FYI, it's a bit more subtle than that. I think that inner maps which you look up from a nested map can have a name. But the "template" map that we create from MapSpec.InnerMap doesn't benefit from the name since it's ephemeral. If this function took a `*btf.Struct` you could just as well pass `Struct.name()` here. No need to change anything though.

Reply:
> I think that inner maps which you look up from a nested map can have a name.
Interesting, do you have an example? Aren't these just plain (outer) map lookups? :thinking:

Reply:
Here is what happens for map in map:

```
template = create_map(ARRAY, ...)
maps = create_map(ARRAY_OF_MAPS, template, ...)
close(template)

somemap = create_map(ARRAY, "my_name", ...)
insert(maps, 0, somemap)

// later

somemap = lookup(maps, 0)
```

Maps aren't created on the fly, you still have to create them yourself and insert them.

Reply:
Of course, now I understand.. libbpf names these `<outer_map>.inner`, will do the same here.